File: src/ruby/app/validators/name_validator.rb
class NameValidator
  def initialize(name)
    @name = name
  end

  def is_valid?
    has_only_letters_or_dashes &&
    has_more_than_two_words &&
    has_only_one_space_between_words &&
    has_capitalized_words &&
    has_first_letter_after_dash_capitalized &&
    has_only_first_letter_capitalized
  end

  private

  attr_accessor :name

  def splitted_name
    name.split(' ')
  end

  def has_only_letters_or_dashes
    return false unless name =~ /\A[a-z\s\-]*\Z/i
    true
  end

  def has_more_than_two_words
    splitted_name.size >= 2
  end

  def has_capitalized_words
    splitted_name.each do |name|
      return false unless name[0] =~ /[[:upper:]]/
    end
    true
  end

  def has_only_one_space_between_words
    return false if name =~ /\s{2,}/
    true
  end

  def has_first_letter_after_dash_capitalized
    return false if name =~ /\-[a-z]/
    true
  end

  def has_only_first_letter_capitalized
    splitted_name.each do |name|
      return false unless name[1..-1] =~ /\A[a-z]*(\-|\z)/
    end
    true
  end
end

File: src/python/validators/name_validator.py
import re

class NameValidator:
    def is_valid(self, name):
        namePattern = "^[A-Z][a-z]{1,}$"
        if " " in name:
            name_words = name.split(" ")
            for name_word in name_words:
                if "-" in name_word:
                    words_with_dash = name_word.split("-")
                    for word_with_dash in words_with_dash:
                        if not (re.match(namePattern, word_with_dash)):                             
                            return False
                else:
                    if not (re.match(namePattern, name_word)):          
                        return False
            return True    
        else:
            return False
        

